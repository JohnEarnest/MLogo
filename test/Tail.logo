# Tail
#
# verify that tail-calls work and are
# optimized correctly.

# short-cut tail-call which should fit on the stack:
to small :x
	if equal? :x 50 [ print 'complete stop ]
	small sum :x 1
end
small 0

# returning tail-call which should fit on the stack:
to enough :x
	if equal? :x 50 [ output 'complete ]
	output enough sum :x 1
end
print enough 0

# short-cut tail-call which will exceed the callstack:
to big :x
	if equal? :x 2000 [ print 'complete! stop ]
	big sum :x 1
end
big 0

# returning tail-call which will exceed the callstack:
to notenough :x
	if equal? :x 2000 [ output 'complete! ]
	output notenough sum :x 1
end
print notenough 0

# a different way to terminate a tail call:
to evenmore :x
	if greater? :x 0 [ evenmore difference :x 1 stop ]
	print 'evenmore!
end
evenmore 2000