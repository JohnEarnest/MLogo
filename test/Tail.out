'complete
'complete
'complete!
'complete!
'evenmore!
